src/WinRT.Runtime/ObjectReference.cs

```diff
@@ -333,47 +324,39 @@ internal unsafe bool IsReferenceToManagedObject
         {
             get
             {
-                return VftblIUnknown.Equals(IUnknownVftbl.AbiToProjectionVftbl);
+                ThrowIfDisposed();
+                return (**(IUnknownVftbl**)ThisPtr).Equals(IUnknownVftbl.AbiToProjectionVftbl);
```

Comment:
More of a general thing I've noticed in CsWinRT, but all of these methods don't seem thread safe at all. You could have a thread concurrently dispose an instance right after another thread got past the disposal check, but before it actually invoked the native method with the obtained this pointer. Would that not crash? Is this just explicitly not supported? Essentially, consider this sequence of events:
- Thread A calls `ProjectedType.Foo()`
- A calls `ThisPtr`, the dispose check passes and it gets the `IntPtr` value (not incremented)
- Thread B calls `Dispose()`, which releases the object
- A now uses that `IntPtr` to invoke some function pointer
- 💥

Is there some guarantee that makes this not actually possible?

Reply:
This is certainly possible. The ObjectReference owns the ThisPtr and if you disposed that while still in use then all kinds of things would go wrong. This is the same as calling SafeHandle.DangerousGetHandle().

Reply:
Out of scope for this PR.